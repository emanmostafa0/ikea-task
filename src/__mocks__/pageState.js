export const products = [
    { 
        name: "test1",
        id: 1,
        quantity: 1,
        articles: [
            {id: 11, amountRequired: 2}
        ]
      
    },
    { 
        name: "test2",
        id: 2,
        quantity: 1,
        articles: [
            {id: 11, amountRequired: 2}
        ]
    }
];

export const articles = [
    { 
        name: "article1",
        id: 11,
        amountInStock: 2
      
    },
    { 
        name: "article2",
        id: 22,
        amountInStock: 2
    }
];