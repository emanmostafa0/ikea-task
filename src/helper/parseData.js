export const parseProductArticles = (products, articles) => {
    products?.length && products?.forEach(product => {
        let minQuantity = product?.articles?.length? Number.MAX_VALUE : 0;
        product.articles = product?.articles?.map(article => {
            const selectedArticle = articles?.find(aricleItem => aricleItem.id === article.id);
            
            const quantity = selectedArticle?.amountInStock ? 
                Math.floor(selectedArticle?.amountInStock/article.amountRequired): 0;
            minQuantity= Math.min(quantity, minQuantity);
            return {...article, name: selectedArticle?.name};
        });
        product.quantity = minQuantity;
    });
    return products;
};

export const updateArticles = (articles, updatedArticles) => {
    return articles?.length && articles?.map(article => {
        const selectedArticle = updatedArticles?.length &&
             updatedArticles?.find(updatedArticle => updatedArticle.id === article.id);
        if(selectedArticle)
            return {...article, amountInStock: selectedArticle?.amountInStock};
        else return article;
    }) || {};
}; 

export const updateProductArticles = (products, articles, updatedArticles) => {
    const newArticles = updateArticles(articles, updatedArticles);
    return products?.length && products?.map(product => {
        let minQuantity = product?.articles?.length? Number.MAX_VALUE : 0;
        product.articles?.forEach(article => {
            const selectedArticle = newArticles?.length && 
                newArticles?.find(aricleItem => aricleItem.id === article.id);
            const quantity = Math.floor(selectedArticle?.amountInStock/article.amountRequired);
            minQuantity= Math.min(quantity, minQuantity);
        });
        return {...product, quantity: minQuantity};
    }) || {};
    
};

