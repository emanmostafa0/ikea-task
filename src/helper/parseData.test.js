import { parseProductArticles, updateProductArticles } from "./parseData";
import {products, articles} from '../__mocks__/pageState';

describe("paseData", () => {
    it("parse products", () => {
        let productsCopiedTest1 = products.map(product => {
            // eslint-disable-next-line no-unused-vars
            const {quantity, ...newProduct} = product;
            return newProduct;
        });
       
        const parsedProduct = parseProductArticles(productsCopiedTest1, articles);
        expect(parsedProduct).toMatchSnapshot([
            { name: 'test1', id: 1, articles: [  {id: 11, amountRequired: 2} ], quantity: 1 },
            { name: 'test2', id: 2, articles: [ {id: 11, amountRequired: 2} ], quantity: 1 }
        ]);
    });

    it("update products", () => {
        const updatesArticles = [
            { 
                name: "article1",
                id: 11,
                amountInStock: 1
              
            }
        ];
        const parsedProduct =updateProductArticles (products, articles, updatesArticles);
        expect(parsedProduct).toMatchSnapshot([
            { name: 'test1', id: 1, articles: [  {id: 11, amountRequired: 2} ], quantity: 0 },
            { name: 'test2', id: 2, articles: [ {id: 11, amountRequired: 2} ], quantity: 0 }
        ]);
    });


   
});