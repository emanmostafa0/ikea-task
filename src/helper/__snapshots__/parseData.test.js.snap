// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`paseData parse products 1`] = `
Array [
  Object {
    "articles": Array [
      Object {
        "amountRequired": 2,
        "id": 11,
        "name": "article1",
      },
    ],
    "id": 1,
    "name": "test1",
    "quantity": 1,
  },
  Object {
    "articles": Array [
      Object {
        "amountRequired": 2,
        "id": 11,
        "name": "article1",
      },
    ],
    "id": 2,
    "name": "test2",
    "quantity": 1,
  },
]
`;

exports[`paseData pase products 1`] = `
Array [
  Object {
    "articles": Array [
      Object {
        "amountRequired": 2,
        "id": 11,
        "name": "article1",
      },
    ],
    "id": 1,
    "name": "test1",
    "quantity": 1,
  },
  Object {
    "articles": Array [
      Object {
        "amountRequired": 2,
        "id": 11,
        "name": "article1",
      },
    ],
    "id": 2,
    "name": "test2",
    "quantity": 1,
  },
]
`;

exports[`paseData update products 1`] = `
Array [
  Object {
    "articles": Array [
      Object {
        "amountRequired": 2,
        "id": 11,
      },
    ],
    "id": 1,
    "name": "test1",
    "quantity": 0,
  },
  Object {
    "articles": Array [
      Object {
        "amountRequired": 2,
        "id": 11,
      },
    ],
    "id": 2,
    "name": "test2",
    "quantity": 0,
  },
]
`;
